Fixing up the Double/Single parsing code to be correct

tannergooding opened the pull request:
This cleans up the `System.Double` and `System.Single` parsing code to be correct. It is a port of the Roslyn "RealParser" code with a few modifications:
* It was updated to use our `Number.BigInteger` structure, rather than `System.Numerics.BigInteger`
* It was updated to have a fast-path for sequences up to 15 digits and exponents up to +/-22
  * With plans for a future PR to expand this to 19 digits and exponents up to +/-27 (using 80-bit extended precision floating-point arithmetic)

For numbers that hit the fast-path, we are no slower than the current implementation and up to 30% faster. For numbers that have to fallback to the slow-path, the worst case I saw was 500% slower (for `double.Epsilon`).
* The slowdown was expected for the `slow-path`, as you must consider and construct a `BigInteger` containing up to 768 digits (but no more digits than the length of the input string). You must then do various arithmetic operations (including division) to compute the correct and precisely rounded result.
  * There are a likely a few places we could improve the algorithm, if and as needed (called out throughout the PR).
* It is additionally worth noting that, the numbers that fall into the slow-path are not expected to be common input cases.
  * After the number has been parsed from it's exact string, you can round-trip it (to string, and back to the nearest representable double) using at most 17 digits. As such, most inputs are expected to have 17 or less significant digits
  * Due to the distribution of floating-point numbers, over half of the unique values lie between -1 and +1 (the delta between each unique value changes at every power of 2). As such, most inputs are expected to fall in this range
  * One could also speculate that the magnitude of most inputs will be between 10^27 and 10^-27, given the range and a looking at a list of special values that fall into and outside this range: https://en.wikipedia.org/wiki/Orders_of_magnitude_(numbers)

This should resolve:
* https://github.com/dotnet/coreclr/issues/1316
* https://github.com/dotnet/coreclr/issues/17467
* https://github.com/dotnet/coreclr/issues/16783

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBitsRoslyn.cs:
```diff
@@ -0,0 +1,499 @@
+{
+    internal unsafe partial class Number
+    {
+        // The below implementation comes from the Roslyn RealParser: https://github.com/dotnet/roslyn/blob/82b2f694b26e5933ff8dd0856a2cd775273ac138/src/Compilers/Core/Portable/RealParser.cs
```
This needs to be properly attributed (etc) before we can merge.

tannergooding commented:
Marked as [WIP] until I can get Roslyn attributed correctly and until I can get the "fast-path" up for both double and single (so we don't regress perf for the common case). For the "fast-path" we can use algorithm similar to the previous `NumberToDouble` implementation, where we rely on floating-point multiplication/division to handle inputs that contain up to 15 digits (both the mantissa and the scale need to meet this requirement).

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -198,6 +406,217 @@ public static int Compare(ref BigInteger lhs, ref BigInteger rhs)
 
+            public static uint CountSignificantBits(uint value)
+            {
+                return (value != 0) ? (1 + LogBase2(value)) : 0;
```
There are a few places, like this, where HWIntrinsics could probably be used in a later PR.

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -198,6 +406,217 @@ public static int Compare(ref BigInteger lhs, ref BigInteger rhs)
+
+            public static void DivRem(ref BigInteger lhs, ref BigInteger rhs, out BigInteger quo, out BigInteger rem)
+            {
+                // This is modified from the CoreFX BigIntegerCalculator.DivRem.cs implementation:
```
Probably some room for improvements here. We only care about the lower 64-bits of `quo`, for example; and `rem` only matters if it is zero or non-zero.

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -410,12 +844,12 @@ public static void Multiply(ref BigInteger lhs, ref BigInteger rhs, ref BigInteg
             }
 
-            public static void Pow10(uint exponent, ref BigInteger result)
+            public static void Pow10(uint exponent, out BigInteger result)
```
Probably some room for improvement here as well, calculating some of the larger powers of 10 can be "expensive".

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBitsRoslyn.cs:
```diff
@@ -0,0 +1,499 @@
+
+            BigInteger.DivRem(ref fractionalNumerator, ref fractionalDenominator, out BigInteger bigFractionalMantissa, out BigInteger fractionalRemainder);
+            ulong fractionalMantissa = bigFractionalMantissa.ToUInt64();
+            bool hasZeroTail = fractionalRemainder.IsZero();
```
This, and other cases of `hasZeroTail` aren't quite correct yet. `NumberBuffer` needs to carry an additional parameter that tracks whether any digits that were parsed, but not stored in the buffer (so digits past 113 for single and past 768 for double), were non-zero to ensure that rounding is always correct/accurate.

tannergooding replied:
Resolved.

tannergooding commented:
CoreFX test failures are expected as `-0.0` is now parsed correctly.
Ran the Roslyn `RealParser` tests and validated that all special inputs are correctly handled
Numbers (before adding a fast path) show inputs with small exponents (like `"1"`) are 2x slower and inputs with large exponents (such as `"4.94065645841247E-324"`) are 5x slower.
* These are numbers from the CoreFX perf tests

tannergooding commented:
A simple fast-path is up, which is marginally faster than the previous implementation for inputs of `<= 15` digits and where the exponent is `<= 22`. For values outside that range, the fallback implementation comes into play and it is slower. Still working on some of the other fixes.

tannergooding commented:
CC. @danmosemsft, @jkotas, @eerhardt as a preliminary FYI.

jkotas commented on src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems:
```diff
@@ -294,7 +294,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Number.Grisu3.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Number.NumberBuffer.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Number.NumberToDouble.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Number.NumberToFloatingPointBitsRoslyn.cs" />
```
This should have more self-describing name than "Roslyn".

tannergooding replied:
`NumberToFloatingPointBits` is probably the most self-describing name. I can remove the `Roslyn` moniker from the end.

tannergooding replied:
Fixed.

danmosemsft replied:
Did you mean to also rename `NumberToFloatingPointBitsRoslyn` method?

tannergooding replied:
Yes, will fix.

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.Parsing.cs:
```diff
@@ -1797,22 +1815,45 @@ private static void ThrowOverflowOrFormatException(bool overflow, string overflo
+            ulong bits = NumberToFloatingPointBitsRoslyn(ref number, in FloatingPointInfo.Double);
+            double result = BitConverter.Int64BitsToDouble((long)(bits));
+
+            if (!double.IsFinite(result))
```
Our current behavior is to return `false` and `0` for overflow. However, the correct behavior is to return `Infinity` with the appropriate sign.

Fixing this would fall into the realm of breaking changes and we would need to determine whether this counts as a parsing failure (which would return `false` from `TryParse` and would throw for `Parse`).

tannergooding replied:
Under the default IEEE rules, this would signal the `overflow exception` (if enabled); However, we explicitly disable and do not support IEEE exception handling; which makes the current behavior a bit of an oddity.

tannergooding replied:
CC. @jkotas, do you have an opinion here on the correct behavior?

I believe the options are:
* Keep the current behavior and return false, setting `result` to 0
* Return false, but set the result to `Infinity` with the appropriate sign
  * This would throw for `Parse`, but for `TryParse` would allow interested users to differentiate between "Format" and "Overflow"; it should not cause any difference in behavior for existing code paths
* Return true and set the result to `Infinity` with the appropriate sign
  * Both `Parse` and `TryParse` would work and this is arguably the most correct. However, some code may not know to handle the overflow case correctly

tannergooding replied:
I've selected to `return true` and set the result to `Infinity` with the appropriate sign, for now. It is the most correct behavior and is what other languages/frameworks do when IEEE exception handling is disabled.

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.Parsing.cs:
```diff
@@ -1576,18 +1592,11 @@ private static unsafe bool TryNumberToDecimal(ref NumberBuffer number, ref decim
             {
-                // If the next digit is 5, round up if the number is odd or any following digit is non-zero
-                if (c == '5' && (low64 & 1) == 0)
+                if ((c == '5') && ((low64 & 1) == 0) && !number.HasNonZeroTail)
```
CC. @pentp on https://github.com/dotnet/coreclr/pull/20707/commits/f9f0034fa516c085bb538b7441c4756e5e86d6ee, since it touches `System.Decimal`.

tannergooding commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:
```diff
@@ -0,0 +1,580 @@
+            uint fractionalLastIndex = totalDigits;
+            uint fractionalDigitsPresent = fractionalLastIndex - fractionalFirstIndex;
+
+            // When the number of significant digits is less than or equal to 15 and the
```
For `System.Single`, this covers a very good range of possible inputs. You only need at most 9 digits to guarantee round-tripping and the worst case is `1.4 * 10^-45`. This means that, for an expected range of inputs, we will likely always be faster than previous, and there will be some very small/large numbers that are slower.

For `System.Double`, this likewise covers a good range of possible inputs. However, you need at most 17-digits to guarantee round-tripping so there are likely some round-trippable inputs that will always be slower (but that will be correct now). There will also be a much broader range of values that will be slower, since the worst case is `4.9 * 10^-324`, but these fall outside the range of what I would expect to be "normal" inputs (IIRC, approx half of all unique values fall between `-1.0 and 1.0`, due to the representation, and for those, I think we can assume the magnitude of most values will be larger than `1e-22` and less than `1e22`).
* There may be some other tricks we can do to speed up this more for things like the 17 digit case. For example: the "golden standard" by David M. Gay, uses a software implementation of an "extended-precision" float with 96-bits to cover some of these inputs (we would need to do all the appropriate license checks before using it, however, other languages like Java use a variation of it so it probably won't be too much of a concern).

tannergooding replied:
If I did my math right, we can have a couple other fast paths to cover the following additional ranges. Doing so would provide code similar to the previous implementation, which was providing a software implementation for 80-bit extended-precision multiply; however, we would also need to provide a software-based division implementation, since negative exponents are not exactly representable.

* System.Single (8-bit exponent, 24-bit mantissa)
  * 7 Digits, "fast exponent" up to 10
* System.Double (11-bit exponent, 53-bit mantissa)
  * 15 Digits, "fast exponent" up to 22
* Extended-Precision Double (15-bit exponent, 64-bit mantissa)
  * 19 Digits, "fast exponent" up to 27
* Quad-Precision (15-bit exponent, 113-bit mantissa)
  * 34 Digits, "fast exponent" up to 38

Supporting a basic software implementation for "extended-precision double" is more trivial since it basically just relies on things we have existing hardware support for (such as `64*64=128`). This would also cover most doubles that are of "round-trippable" length (17 digits).

Supporting a basic software implementation for "quad-precision" is more complex, but may still be worthwhile, since it would effectively cover the entire range of `System.Single` "normal" inputs (for `System.Single`, only subnormal inputs have an exponent that is greater than 38).

tannergooding replied:
It now appropriately uses `System.Single` or `System.Double` arithmetic, where applicable.

It also uses `Extended-Precision` arithmetic for the multiplication case. For the division case (which is arguably more common), the implementation isn't as trivial. Berkeley has an existing software-implementation (available under the BSD 3 Clause) for most of the floating-point functions (for binary16, binary32, binary64, binary80, and binary128), which would be nice if we could use here (rather than needing to port our own): https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/COPYING.txt

danmosemsft commented on src/System.Private.CoreLib/shared/System/Number.BigInteger.cs:
```diff
@@ -27,14 +45,16 @@ internal static partial class Number
+                116,        // 10^1024
             };
 
             private static readonly uint[] s_Pow10BigNumTable = new uint[]
```
Is it feasible/worthwhile to have debug-only code that validates the entries in this table don't contain a mistake? I didn't see any

tannergooding replied:
Yes, we have that elsewhere.

danmosemsft commented on src/System.Private.CoreLib/shared/System/Number.NumberBuffer.cs:
```diff
@@ -10,7 +11,7 @@ namespace System
     {
         //  We need 1 additional byte, per length, for the terminating null
-        private const int DecimalNumberBufferLength = 50 + 1;
+        private const int DecimalNumberBufferLength = 30 + 1;  // 29 for the longest input + 1 for rounding
```
Possibly write these as 
`private const int DecimalNumberBufferLength = 29 +1 + 1;  // 29 for the longest input + 1 for rounding`
because the +1 you see is not the +1 for rounding...

tannergooding replied:
Fixed.

danmosemsft commented:
> Try decimal.Parse("1234567890123456789012345.678456") - it should round to .6785, but at the moment results in .6784

Missing test?

pentp commented:
> 
> 
> > Try decimal.Parse("1234567890123456789012345.678456") - it should round to .6785, but at the moment results in .6784
> 
> Missing test?

No, this value came out from a test, just pointed it out as a quick example.

tannergooding commented:
The CoreFX failures are here (linked since they should now be disabled): https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_windows_nt_corefx_innerloop_prtest/3912/#showFailuresLink

There were 159 in total and were caused by:
* `"-0"` now parsing as `-0`
* "Overflow" now returning the appropriately signed Infinity, rather than 0 and rather than throwing
* The UTF8 Span Parser not being in sync with the new changes
* Four decimal parse tests that found a bug and were fixed, they were not disabled

tannergooding commented:
A number of the CoreFX failures are unrelated and are showing up on other PRs.

pentp commented on src/System.Private.CoreLib/shared/System/Number.Parsing.cs:
```diff
@@ -1576,18 +1592,20 @@ private static unsafe bool TryNumberToDecimal(ref NumberBuffer number, ref decim
+
+                    // At this point we should either be at the end of the buffer, or just
+                    // have a single rounding digit left, and the next should be the end
+                    Debug.Assert((c == 0) || (*++p == 0));
```
This assert modifies the value of `p`.

jkotas commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:
```diff
@@ -0,0 +1,762 @@
+            return Multiply32x32To64(au, bu) + du + eu;
+        }
+
+        private static ulong NumberToFloatingPointBits(ref NumberBuffer number, in FloatingPointInfo info)
```
This method is ~300 lines long now - I think it is close to hitting the various internal JIT limits, if it is not over them already. Would it make sense to split the slow path into a separate method at least?

tannergooding replied:
Yes, I'll split it.

tannergooding commented:
I believe the only remaining items are:
* Ensure Roslyn is properly attributed (waiting for response on e-mail thread)
* Optionally add some debug validation code for the Extended-Precision and BigInteger Pow10 tables
* Update the extended-precision fast-path to cover division (waiting for response on e-mail thread, we should do this so that some 17-digit sequences are covered by the fast path)
* Optionally add a fast-path using `binary128` arithmetic (this can wait for a separate PR, and would be pending the e-mail response listed above)

tannergooding commented:
I believe this is ready for final review.

After some discussion with @danmosemsft, it sounds like we want to log issues for some of the remaining optional work and look into them at some point in the future (such as if we get feedback about other inputs, outside the currently expected range).

tannergooding commented:
CC. @gafter since you logged https://github.com/dotnet/coreclr/issues/1316. I ran the entire Roslyn "RealParser" test suite (modified to call `double.TryParse` and `float.TryParse`) and all cases passed.

tannergooding commented:
Fixed a couple bugs caught by the test base provided by @cyberphone: https://github.com/dotnet/coreclr/issues/17467#issuecomment-384932600
* We were inserting the `digitEnd` (null character) after the last non-zero digit
* The 80-bit mantissa path has an issue with its mantissa normalization logic that was causing off-by-one
* I also fixed-up the parsing case where we should check for `Infinity` and `NaN` case-insensitively.

Now, the entire Roslyn RealParser suite and the entire ES6 test suite (covering 100M inputs) passes validation.

danmosemsft commented:
Shoukd we pick up some of the Roslyn tests as well?

eerhardt commented on THIRD-PARTY-NOTICES.TXT:
```diff
@@ -249,3 +249,208 @@ LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+      (a) You must give any other recipients of the Work or
+          Derivative Works a copy of this License; and
+
+      (b) You must cause any modified files to carry prominent notices
```
Are you following this condition in the files you modified?

tannergooding replied:
No, I lost that in the latest commit. Will fix.

gafter replied:
Since this is Microsoft-written code, we can re-release it under any license we want.

jkotas commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:
```diff
@@ -0,0 +1,726 @@
+// See https://github.com/dotnet/roslyn/blob/82b2f694b26e5933ff8dd0856a2cd775273ac138/src/Compilers/Core/Portable/RealParser.cs
+
+// This file is a port of the original code and has been modified to have certain performance improvements
+// and to better fit in with the requirements of CoreCLR and existing in System.Private.Corelib.
```
and existing -> and existing code ?

tannergooding replied:
Fixed.

gafter commented:
I haven't looked at the code yet, but I'm wondering... on what basis can we assert that the fast path is "correct"?

tannergooding commented:
> on what basis can we assert that the fast path is "correct"?

IEEE 754:2008 5.1 Operations
> Each of the computational operations that return a
numeric result specified by this standard shall be performed as if it first produced an intermediate result
correct to infinite precision and with unbounded range, and then rounded that intermediate result, if
necessary, to fit in the destination’s format

Additionally, any integer up to 2^53 is exactly representable (which is 15 digits, if the most significant digit must be allowed to go up to `9`) as well as powers of 10 up to 22. This guarantees that, for a platform correctly implementing IEEE division and multiplication operations, we will get a correctly rounded result.

tannergooding commented:
`IEEE 754:1985 5. Operations` lists the same requirement for computation of results for multiplication/division.

gafter commented:
How do we reconcile this
> any integer up to 2^53 is exactly representable (which is 15 digits, if the most significant digit must be allowed to go up to 9) as well as powers of 10 up to 22

With this
> It was updated to have a fast-path for sequences up to 19 digits and exponents up to +/-27

How do we know that it is correct when extended to 19 digits and exponents up to +/-27?

tannergooding commented:
> > It was updated to have a fast-path for sequences up to 19 digits and exponents up to +/-27
>
> How do we know that it is correct when extended to 19 digits and exponents up to +/-27?

Ah, I'll correct the comment. That was for the 80-bit Extended Precision code-path, which was removed due to a mismatch between the normalization logic done as part of software multiplication and the normalization logic that `AssembleFloatingPointBits` did (it was basically getting normalized twice).
* 80-bit extended precision floating-point math allows up to 19 digits and powers of 10 up to 27

tannergooding commented:
> Can you add the Roslyn tests? It covers edge cases that are the easiest to get wrong.

We normally add the relevant test cases to CoreFX, which requires this to be merged first. CC. @jkotas

tannergooding commented:
@danmosemsft, @ViktorHofer, @eerhardt: Could one of you take a look here. I'd like to get another set of eyes on this before merging.

Also CC @jkotas to see if he has any more feedback.

tannergooding commented:
Rebased to resolve the merge conflict in `tests/CoreFX/CoreFX.issues.json`

ViktorHofer commented:
Will take a look later today.

jkotas commented:
> Also CC @jkotas to see if he has any more feedback.

As mentioned above and offline, I think it would be better to avoid all the licensing glue and just submit this under the regular one since this is all Microsoft written code and thus we can do it just fine.

eerhardt commented on src/System.Private.CoreLib/shared/System/Number.Parsing.cs:
```diff
@@ -1797,22 +1833,33 @@ private static void ThrowOverflowOrFormatException(bool overflow, string overflo
+            return true;
+        }
+
+        private static bool TryNumberToSingle(ref NumberBuffer number, out float value)
```
When does this method ever return something other than `true`? If it doesn't, maybe it shouldn't be a `Try` method then.

eerhardt replied:
Same for the above method.

---
In reply to: [231663340](https://github.com/dotnet/coreclr/pull/20707#discussion_r231663340) [](ancestors = 231663340)

tannergooding replied:
Will fixup. It no longer returns `false` since `Overflow` will now return the appropriately signed `Infinity`, rather than false.

tannergooding replied:
Fixed.

eerhardt commented on src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs:
```diff
@@ -0,0 +1,627 @@
+            bool roundBit = (value & roundBitMask) != 0;
+            bool hasTailBits = !hasZeroTail || (value & extraBitsMask) != 0;
+
+            return (value >> shift) + (ShouldRoundUp(lsbBit: lsbBit, roundBit: roundBit, hasTailBits: hasTailBits) ? 1UL : 0);
```
(nit) the method call here is a bit redundant in the argument passing, which makes it hard to read IMO:

`ShouldRoundUp(lsbBit: lsbBit, roundBit: roundBit, hasTailBits: hasTailBits)` =>
`ShouldRoundUp(lsbBit, roundBit, hasTailBits)`

tannergooding replied:
Fixed.

tannergooding commented:
@jkotas, latest changes have remove the licensing "glue" and re-released the code under MIT.

tannergooding commented:
Most of the jobs stopped due to disconnection.

pentp commented:
The CI system seems to have serious problems at the moment 😕

danmosemsft commented:
Yay @tannergooding !

danmosemsft commented:
> Edit: my comment triggered all the tests now....

@mmitche I have seen this too -- occasionally, just ocmmenting seems to kick off CI again. Do you know why?

mmitche commented:
It can happen on old PRs if Jenkins loses track of the current state of what it has tested.
